### ISSUE TITLE
Checking JSX elements ignores `JSX.IntrinsicElements` declared using template literal index signatures

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#55240]
# Bug Report

### 🔎 Search Terms

JSX IntrinsicElements template literal types index signatures

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=1&ts=5.2.0-dev.20230802#code/CYUwxgNghgTiAEYD2A7AzgF3gJRFMGAXPAK4oDWKSA7igFB2iSwIpQC2IaADvggFIBlABrwA3nXhT4IAB7ckMLAEsUGEDABmfeAEk1MVWmVgAohBCc1acZOn2A2iAvsAchxDEABuswASMUxDFABzAF8vAF1iMTC7KTi4ugBuAB5fDE0kJHgAegA+OiA)

### 💻 Code

```ts
declare const React: unknown

declare namespace JSX {
    export interface IntrinsicElements {
        [elemName: `test${string}`]: {}
    }
}

;<testfoo /> // Property 'testfoo' does not exist on type 'JSX.IntrinsicElements'.(2339)
```

### 🙁 Actual behavior

An error is reported claiming that `testfoo` does not exist on type `JSX.IntrinsicElements`

### 🙂 Expected behavior

I'd expect that index signature to permit this JSX element.
